### PR TITLE
Code metrics moving and redesign

### PR DESCRIPTION
**Pull request by Vogel612**
The following things are done:

- Move `Rubberduck.Inspections` to `Rubberduck.CodeAnalysis`
- Move CodeMetrics from `Rubberduck.Core` into `Rubberduck.CodeAnalysis`
- Overhaul the design for CodeMetrics' internal workings
- Reimplement the existing CodeMetrics with the new design (CC, LineCount and Nesting done)

TBD:

- UI. A metric ton of UI work is necessary here. As of now the UI is broken, help wanted.

See #3862 for a bit of background, as well as #403 and #3522's review comments for why this is done

### CODE REVIEW

File: Rubberduck.CodeAnalysis/CodeMetrics/CodeMetricBase.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using System.Collections.Generic;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.CodeAnalysis.CodeMetrics
+{
+    internal abstract class CodeMetricResultBase : ICodeMetricResult
+    {
+        public CodeMetricResultBase(Declaration declaration, CodeMetric metric)
```
Comment (bclothier) [inline Rubberduck.CodeAnalysis/CodeMetrics/CodeMetricBase.cs]: isn't it weird to have a `internal` class that contains a `public` ctor?

File: Rubberduck.CodeAnalysis/CodeMetrics/LineCountMetric.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        public override void Reset()
+        {
+            base.Reset();
+            workingValue = 0;
+        }
+
+        public override IEnumerable<ICodeMetricResult> Results() => new[] { new LineCountModuleMetricResult(_finder.ModuleDeclaration(_qmn), ownerReference, workingValue) };
+
+        public override void EnterEndOfLine([NotNull]  VBAParser.EndOfLineContext _) => workingValue++;
```
Comment (bclothier) [inline Rubberduck.CodeAnalysis/CodeMetrics/LineCountMetric.cs]: Do we wanna to count blank and comment lines? Or better, enhance to return a `struct` with counts of each types `Declarations`, `Blank`, `Comment line`, and `Module body line` with a total?

File: Rubberduck.Main/Root/RubberduckIoCInstaller.cs [hunk 3]
```diff
@@ -146,13 +147,41 @@ public void Install(IWindsorContainer container, IConfigurationStore store)
...
+            var metricImplementations = new List<CodeMetric>();
+            foreach (var assembly in assembliesToRegister)
+            {
+                metricImplementations.AddRange(assembly.GetTypes()
+                    .Where(t => t.BaseType == typeof(CodeMetric))
+                    .Select(t =>
+                    {
+                        try
+                        {
+                            return (CodeMetric)Activator.CreateInstance(t, new object[] { });
```
Comment (Vogel612) [inline Rubberduck.Main/Root/RubberduckIoCInstaller.cs]: I kinda want a cleaner way than instantiating these at registration time

File: Rubberduck.Main/Root/RubberduckIoCInstaller.cs [hunk 4]
```diff
@@ -146,13 +147,41 @@ public void Install(IWindsorContainer container, IConfigurationStore store)
...
+                        catch (Exception e)
+                        {
+                            // FIXME: log instantiation failure
+                            return null;
+                        }
+                    }).Where(m => m != null));
+            }
+            container.Register(Component.For<ICodeMetricsAnalyst>()
+                .ImplementedBy<CodeMetricsAnalyst>()
+                .DependsOn(Dependency.OnValue<IEnumerable<CodeMetric>>(metricImplementations))
```
Comment (Vogel612) [inline Rubberduck.Main/Root/RubberduckIoCInstaller.cs]: a rare case of explicit dependency configuration. Is that okay?

File: Rubberduck.Main/Root/RubberduckIoCInstaller.cs (hunk 3 above)
Comment (MDoerner) [reply Rubberduck.Main/Root/RubberduckIoCInstaller.cs]: Why do you not just register all the code metrics with `CodeMetric`? Then the collection of all of them will be registered with `IEnumerable<CodeMetric>`.

Comment (Vogel612) [reply Rubberduck.Main/Root/RubberduckIoCInstaller.cs]: I tried that before, it didn't work :/

File: Rubberduck.CodeAnalysis/Inspections/Inspector.cs [hunk 5]
```diff
@@ -53,7 +55,6 @@ private void UpdateInspectionSeverity(Configuration config)
                         if (inspection.Name == setting.Name)
                         {
                             inspection.Severity = setting.Severity;
-                            continue;
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Inspector.cs]: Shouldn't this be a `break`?

File: Rubberduck.Core/CodeAnalysis/CodeMetrics/CodeMetricsViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+        private Dictionary<Declaration, List<ICodeMetricResult>> resultsByDeclaration;
+
+        private CodeExplorerItemViewModel _selectedItem;
+        public CodeExplorerItemViewModel SelectedItem
+        {
+            get => _selectedItem;
+            set
+            {
+                _selectedItem = value;
```
Comment (retailcoder) [inline Rubberduck.Core/CodeAnalysis/CodeMetrics/CodeMetricsViewModel.cs]: if `value == _selectedItem`, we're firing extraneous `PropertyChanged` events here.

File: Rubberduck.Core/UI/CodeExplorer/CodeExplorerWindow.cs [hunk 7]
```diff
@@ -8,8 +9,8 @@ namespace Rubberduck.UI.CodeExplorer
     [ExcludeFromCodeCoverage]
     public partial class CodeExplorerWindow : UserControl, IDockableUserControl
     {
-        private const string ClassId = "C5318B59-172F-417C-88E3-B377CDA2D809";
-        string IDockableUserControl.ClassId { get { return ClassId; } }
+        private readonly string RandomGuid = Guid.NewGuid().ToString();
```
Comment (retailcoder) [inline Rubberduck.Core/UI/CodeExplorer/CodeExplorerWindow.cs]: :skull_and_crossbones:  makes me wonder what the purpose of `IDockableUserControl.GuidIdentifier` really is... do we really need this?

File: Rubberduck.Core/UI/Controls/SearchResultWindow.cs [hunk 8]
```diff
@@ -21,10 +22,8 @@ public ISearchResultsWindowViewModel ViewModel
             }
         }
 
-        public string ClassId
-        {
-            get { return "BFD04A86-CACA-4F95-9656-A0BF7D3AE254"; }
-        }
+        private readonly string RandomGuid = Guid.NewGuid().ToString();
```
Comment (retailcoder) [inline Rubberduck.Core/UI/Controls/SearchResultWindow.cs]: If we *do* need a GUID, then why can't it be a `RubberduckGuid` constant?

File: Rubberduck.Core/UI/Converters/DeclarationToMemberSignatureConverter.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            throw new NotImplementedException();
+        }
+
+        private string FullMemberSignature(Declaration member)
+        {
+            var signature = $"{GetMethodType(member)} {member.IdentifierName}({string.Join(", ", GetMemberParameters(member))})";
+
+            return member.AsTypeName == null ? signature : $"{signature} As {member.AsTypeName}";
```
Comment (retailcoder) [inline Rubberduck.Core/UI/Converters/DeclarationToMemberSignatureConverter.cs]: tempted to swap `As` for `{Tokens.As}` here :wink:

File: Rubberduck.Core/UI/Controls/SearchResultWindow.cs (hunk 8 above)
Comment (bclothier) [reply Rubberduck.Core/UI/Controls/SearchResultWindow.cs]: This particular GUID is just used as an unique identifier for an instance created by the `CreateToolWindow`, as [per documentation](https://msdn.microsoft.com/en-us/vba/language-reference-vba/articles/createtoolwindow-method) for `GuidPosition` parameter.

File: Rubberduck.Core/UI/Converters/DeclarationToMemberSignatureConverter.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                        ParamType = p.AsTypeName
+                    })
+                    .ToList();
+            }
+            else
+            {
+                return new List<Parameter>();
+            }
+
+            if (GetMethodType(member) == "Property Get")
```
Comment (retailcoder) [inline Rubberduck.Core/UI/Converters/DeclarationToMemberSignatureConverter.cs]: Feels wrong when the method returns `$"{Tokens.Property} {Tokens.Get}"`, would prefer to have string interpolation here too... but why can we assume `parameters.Count` is >= 1 here? Most `Property Get` members are parameterless... :confused:

File: Rubberduck.Core/UI/Controls/SearchResultWindow.cs (hunk 8 above)
Comment (bclothier) [reply Rubberduck.Core/UI/Controls/SearchResultWindow.cs]: That said, `RandomGuid` really does not convey the reason. I would have preferred a name like `InstanceIdentifier` or something similar.

File: Rubberduck.CodeAnalysis/Inspections/Inspector.cs (hunk 5 above)
Comment (Vogel612) [reply Rubberduck.CodeAnalysis/Inspections/Inspector.cs]: I don't even know where this is coming from ... let me investigate that bit of change

Comment (Vogel612) [reply Rubberduck.CodeAnalysis/Inspections/Inspector.cs]: seems to have been added on next with commit https://github.com/rubberduck-vba/Rubberduck/commit/459d322424721131e6b54b84422a339478dd7889 ... :cough:

**Pull request merged**